clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -217,104 +235,146 @@ private boolean flush(ByteBuffer buf) throws IOException {
-                    //so the selector won't invoke this channel if we don't go through the handshakeFinished here.
-                    if (handshakeStatus != HandshakeStatus.FINISHED) {
-                        if (handshakeStatus == HandshakeStatus.NEED_WRAP) {
-                            key.interestOps(key.interestOps() | SelectionKey.OP_WRITE);
-                        } else if (handshakeStatus == HandshakeStatus.NEED_UNWRAP) {
-                            key.interestOps(key.interestOps() & ~SelectionKey.OP_WRITE);
-                        }
-                        break;
+                } else if (handshakeResult.getStatus() == Status.CLOSED) {
+                    throw new EOFException("SSL handshake status CLOSED during handshake UNWRAP");
+                }
+                log.trace("SSLHandshake NEED_UNWRAP channelId {}, handshakeResult {}, appReadBuffer pos {}, netReadBuffer pos {}, netWriteBuffer pos {}",
+                          channelId, handshakeResult, appReadBuffer.position(), netReadBuffer.position(), netWriteBuffer.position());
+
+                //if handshakeStatus completed than fall-through to finished status.
+                //after handshake is finished there is no data left to read/write in socketChannel.
+                //so the selector won't invoke this channel if we don't go through the handshakeFinished here.
+                if (handshakeStatus != HandshakeStatus.FINISHED) {
+                    if (handshakeStatus == HandshakeStatus.NEED_WRAP) {
+                        key.interestOps(key.interestOps() | SelectionKey.OP_WRITE);
+                    } else if (handshakeStatus == HandshakeStatus.NEED_UNWRAP) {
+                        key.interestOps(key.interestOps() & ~SelectionKey.OP_WRITE);
                     }
-                case FINISHED:
-                    handshakeFinished();
-                    break;
-                case NOT_HANDSHAKING:
-                    handshakeFinished();
                     break;
-                default:
-                    throw new IllegalStateException(String.format("Unexpected status [%s]", handshakeStatus));
-            }
-
-        } catch (SSLException e) {
-            handshakeFailure();
-            throw e;
+                }
+            case FINISHED:
+                handshakeFinished();
+                break;
+            case NOT_HANDSHAKING:
+                handshakeFinished();
+                break;
+            default:
+                throw new IllegalStateException(String.format("Unexpected status [%s]", handshakeStatus));
         }
     }
 
+    /**
+     * SSL exceptions are propagated as authentication failures so that clients can avoid
+     * retries and report the failure. If `flush` is true, exceptions are propagated after
+     * any pending outgoing bytes are flushed to ensure that the peer is notified of the failure.
+     */
+    private void handleSslException(SSLException e, boolean flush) throws IOException {
```

Comment:
Should this be part of `handshakeFailure` since this is only handling `SSLException`s during handshake?

Reply:
Have a single `handshakeFailure` method now which does the code in `handleSslException` as well.